nifi-extension-bundles/nifi-github-bundle/nifi-github-extensions/src/main/java/org/apache/nifi/github/GitHubRepositoryClient.java

```diff
@@ -0,0 +1,404 @@
+     * @throws FlowRegistryException if a non I/O error happens calling GitHub
+     */
+    public List<GHCommit> getCommits(final String path, final String branch) throws IOException, FlowRegistryException {
+        final String resolvedPath = getResolvedPath(path);
+        final String branchRef = BRANCH_REF_PATTERN.formatted(branch);
+        LOGGER.debug("Getting commits for [{}] from branch [{}] in repo [{}]", resolvedPath, branch, repository.getName());
+
+        return execute(() -> {
+            try {
+                final GHRef branchGhRef = repository.getRef(branchRef);
+                return repository.queryCommits()
+                        .path(resolvedPath)
+                        .from(branchGhRef.getObject().getSha())
+                        .pageSize(COMMIT_PAGE_SIZE)
+                        .list()
+                        .toList();
+            } catch (final FileNotFoundException fnf) {
+                throwPathOrBranchNotFound(resolvedPath, branchRef);
+                return null;
+            }
+        });
+    }
+
+    /**
+     * Gets the names of the directories contained within the given directory.
+     *
+     * @param directory the directory to list
+     * @param branch the branch
+     * @return the set of directory names
+     *
+     * @throws IOException if an I/O error happens calling GitHub
+     * @throws FlowRegistryException if a non I/O error happens calling GitHub
+     */
+    public Set<String> getDirectoryNames(final String directory, final String branch) throws IOException, FlowRegistryException {
+        final String resolvedDirectory = getResolvedPath(directory);
+        final String branchRef = BRANCH_REF_PATTERN.formatted(branch);
+        LOGGER.debug("Getting directory names for [{}] from branch [{}] in repo [{}] ", resolvedDirectory, branch, repository.getName());
+
+        return execute(() -> {
+            try {
+                return repository.getDirectoryContent(resolvedDirectory, branchRef).stream()
+                        .filter(GHContent::isDirectory)
+                        .map(GHContent::getName)
+                        .collect(Collectors.toSet());
+            } catch (final FileNotFoundException e) {
+                throwPathOrBranchNotFound(resolvedDirectory, branchRef);
+                return null;
+            }
+        });
+    }
+
+    /**
+     * Gets the names of the directories container within the given directory.
+     *
+     * @param directory the directory to list
+     * @param branch the branch
+     * @return the set of file names
+     *
+     * @throws IOException if an I/O error happens calling GitHub
+     * @throws FlowRegistryException if a non I/O error happens calling GitHub
+     */
+    public Set<String> getFileNames(final String directory, final String branch) throws IOException, FlowRegistryException {
+        final String resolvedDirectory = getResolvedPath(directory);
```

Comment:
It looks to be the same as `getDirectoryNames` expect the filter. It might worth to merge them into a private method and the API methods might call that with different filters. Something like `Set<String> getFilteredContent(final String directory, final String branch, final Filter<GHContent> filter)`